README.adoc

```diff
@@ -1,17 +1,205 @@
+----
+
+You cannot use other methods until the `ULink` is connected. When this happens the `CompletableFuture<UStatus>` returned by connect() will be completed and you also will receive the `onLifecycleChanged(..., true)` callback on your service lifecycle listener. You may query the connected status using these methods:
+
+[,java]
+----
+boolean isDisconnected()
+boolean isConnecting()
+boolean isConnected()
+----
+
+==== Sending a UMessage
+For both, publisher/subscriber or observer (notification) design patterns, a uE should use the `ULink` to send messages to consumers using any method below:
+
+[,java]
+----
+UStatus send(UUri source, UPayload payload, UAttributes attributes)
+UStatus send(UMessage message)
+----
+
+==== Registering a UListener
+In order to start receiving messages, a consumer should register a listener for a topic:
+
+[,java]
+----
+UStatus registerListener(UUri topic, UListener listener)
+----
+*For the publisher/subscriber design pattern*, the precondition for a callback is that the uE needs to subscribe to the topic AND register the listener.
+
+Given the precondition, the callback will be triggered in any of the following cases:
+
+. As soon as listener is registered if there is already a sent message for that topic that is in cache, OR
+. Whenever the producer sends a new message for that topic
+
+*For the notification design pattern*, the only precondition is that uE needs to register the listener.
+Once the listener is registered the callback will be triggered whenever the notification message is sent by the producer.
+
+A consumer can use the same listener for multiple topics, or register different listeners to the same topic.
+
+To unregister a listener from receiving topic messages:
+
+[,java]
+----
+UStatus unregisterListener(UUri topic, UListener listener)
+----
+
+To unregister a listener from all topics:
+
+[,java]
+----
+UStatus unregisterListener(UListener listener)
+----
+
+==== Registering a URpcListener
+A uE with a service role should register a listener for a particular method URI to be notified when request messages are sent against said method.
+
+NOTE: Only one listener is allowed to be registered per a method URI.
+
+[,java]
+----
+UStatus registerRpcListener(UUri methodUri, URpcListener listener)
+----
+
+To stop processing request messages for a specific method URI or all of the, a service uE should unregister the listener:
 
-=== Importing the sdk
- 
-_coming soon_
+[,java]
+----
+UStatus unregisterRpcListener(UUri methodUri, URpcListener listener)
+UStatus unregisterRpcListener(URpcListener listener)
+----
 
-=== Using The Sdk
+==== Invoking an RPC method
+Code generators for uProtocol services defined in proto files primarily utilize the following method. However, clients also have the option to directly use it for invoking RPC methods.
 
-_coming soon_
+[,java]
+----
+CompletableFuture<UPayload> invokeMethod(UUri methodUri, UPayload requestPayload, CallOptions options)
```

Comment:
As discussed offline, can we please use the most abstract type here (i.e. CompletionStage instead of CompletableFuture)?

Reply:
done